csrc/scheduler/normalization_utils.cpp

```diff
@@ -1262,6 +1271,44 @@ bool compileTimeCheck(Fusion* fusion, ScheduleHeuristic schedule_heuristic) {
   return true;
 }
 
+void movePersistentBufferToSmem(
+    Fusion* fusion,
+    const ReductionParams& rparams,
+    const std::vector<TensorView*>& cached_inputs) {
+  // Transfer the persistent buffer tensors to shared memory. These tensors are
+  // housed in smem_persistent_buffers. If a candidate tensor is input, move its
+  // associated cached tensors.
+  if (!rparams.smem_persistent_buffers.empty()) {
+    const auto& persistent_buffers =
+        scheduler_utils::persistentBuffers(fusion).persistent_buffers;
+    auto isSharedMemoryPersistent = [&rparams](const TensorView* lookup_tv) {
+      return std::any_of(
+          rparams.smem_persistent_buffers.begin(),
+          rparams.smem_persistent_buffers.end(),
+          [lookup_tv](const auto* tv) {
+            // can't use `tv->sameAs(lookup_tv)` since the saved tvs in
+            // smem_persistent_buffers are from a cloned fusion.
+            return tv->name() == lookup_tv->name();
+          });
+    };
+    for (auto tv : persistent_buffers) {
+      bool use_smem = isSharedMemoryPersistent(tv);
+      // If project to inputs, the buffer inputs are stored in
+      // [smem_persistent_buffers], after projection, the cached input becomes
+      // the new persistent buffer.
+      if (!use_smem && rparams.project_persistent_buffers &&
+          std::find(cached_inputs.begin(), cached_inputs.end(), tv) !=
+              cached_inputs.end()) {
+        auto input_tv = ir_utils::producerTvsOf(tv).at(0);
+        use_smem = isSharedMemoryPersistent(input_tv);
+      }
```

Comment:
Can you please clarify what's going on here? Why does `isSharedMemoryPersistent` need to be used twice?

Reply:
Revised comments:
```
      // Persistent buffers are categorized into two types:
      // (1) Cached input tensors.
      //     For these, [smem_persistent_buffers] holds the original input
      //     tensors, not the cached.
      // (2) Intermediate tensors: Other tensors used throughout computation.

      // If a buffer is absent from [smem_persistent_buffers], it may be a
      // cached input. In such cases, verify if the original input tensor is
      // stored in [smem_persistent_buffers]. So, we may need to call
      // isSharedMemoryPersistent() twice, one for the buffer iteself and the
      // other for the buffer's input tensor if the buffer is a cached input
      // and it is not in [smem_persistent_buffers].
```
Condition`rparams.project_persistent_buffers` is not required to trigger this additional check since input tensors can be persistent buffer even without projecting to inputs.